### ISSUE TITLE
how to call `spl-token create-account` twice?

### ISSUE THREAD

**Issue by alexiskattan** [solana-labs/solana-program-library#1896]
In the docs, `https://spl.solana.com/token` if you search for `spl-token create-account AQoKYV7tYpTrFZN6P5oUufbQKAUr9mNYGe1TTJC9wajM`, you see that when this function is called twice it produces two different accounts. 

However, when I try and run the program locally, I get, `Error: Account already exists: XXXX`.  Is there some flag to get it to produce an hierarchical address of sorts?

**Comment by CriesofCarrots** [comment]
That looks like a typo in the docs. The `create-account` call under this heading https://spl.solana.com/token#example-transferring-tokens-to-an-explicit-recipient-token-account should have an explicit keypair called out, the one associated with address `CqAxDdBRnawzx9q4PYM3wrybLHBhDZ4P6BTV13WsRJYJ`:

```
spl-token create-account AQoKYV7tYpTrFZN6P5oUufbQKAUr9mNYGe1TTJC9wajM <KEYPAIR>
```

**Comment by CriesofCarrots** [comment]
> Thanks for the help. To clarify here, a keypair is used to generate a token account for a specific token. For a unique keypair, can only ONE token account be created?

Hi @alexiskattan ! In the "associated token account" standard, each wallet (owning keypair) has one token account for each specific token, located at a deterministic address. https://spl.solana.com/token#associated-token-account This way wallets can transfer tokens of any type using just the main wallet address of the recipient.

It is possible for an owner to have additional auxiliary accounts controlled by the same keypair (which is what is created in the command I quoted above), but not all wallet applications may not support transfers to these additional auxiliary accounts, since they are not at the deterministic address. So I would not recommend this route.

> First questions, is there a better way to do this then generating unique addresses? Is there a way to know who/which address sent tokens to an account?

Certainly the transaction will show the address that sent the tokens, so you could inspect the transaction after receipt to ensure it came from User A. But I suppose it could be useful for accounting purposes to receive each payment in a unique account. Not sure I can answer this one better without more details on your use-case.

> Secondly, if only one token account address can be created for a keypair, what is the best way to generate unique addresses for a specific token?

If we're talking about the spl-token-cli, you can pass the `--owner` parameter to use a different keypair/signer. This can be a path to a keypair JSON file, `prompt:` for a [paper wallet](https://docs.solana.com/wallet-guide/paper-wallet), or the [path to a hardware wallet](https://docs.solana.com/wallet-guide/hardware-wallets) like Ledger. The latter two options both support bip44 derivation, so that might make sense for your accounting.

**Comment by alexiskattan** [comment]
Thanks for the info again!  Here is a bit more about the use case.  User A is selling an eBook.  User B comes to User A's shop and says they want to buy the eBook.  User A then generates a unique address for User B.  From what I understand it costs a little bit of SOL to generate the address for a specific token.  

Is there a way to do this without incurring a fee?  The reason is that after User B says they want to buy the eBook they may decide that they don't.    I guess the fee could be charged to User B for making the address but that will make the UX a bit more complicated.  Additionally, if User B is a bad actor, in theory they could drain User A's account by repeatedly saying they want to buy the book, having User A generate and pay for an address, but then not.

**Comment by CriesofCarrots** [comment]
> Is there a way to do this without incurring a fee?

@alexiskattan you are right that creating an account for a specific token always costs a little bit of SOL. However, that SOL is recoverable by the owner by closing the account. That said, it would be a hassle for User A to have to go close thousands of accounts, if you had that process automated and a malicious actor took advantage.
Possibly a more robust solution for you might be an escrow program, where User B sends tokens to a program-owned address, which can be claimed by User A or canceled (reclaimed) by User B. But unfortunately, I don't know of such a deployed program.

> Might there be a base example online for transferring spl tokens from one account to another?

There is an integration-test example: https://github.com/solana-labs/solana-program-library/blob/f519dc75490309b359dfc209af455d735a77b273/token/js/cli/token-test.js#L265

It looks like your issue is the parameters you are passing to the `Token()` construction. I think they should be:
```
const token = new token3.Token(
    connection,                         // connection
    new web3.PublicKey('C2GcearUU9MdByy9ngHAUBYkcxR84WvjCJWBeAu1mMvg'),    // publicKey of token
    TOKEN_PROGRAM_ID,       // programId deployed on chain
    payerAccount,                    // payer
)
```
where `TOKEN_PROGRAM_ID` is the const that ships with the library, `new PublicKey('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA')`, not the one you defined.

That said, please use the #developer-support channel on our Discord instead of github for JS support in the future: https://discord.gg/5MXrE2Fv

**Comment by t-nelson** [comment]
> But unfortunately, I don't know of such a deployed program.

https://github.com/project-serum/anchor/blob/master/examples/cashiers-check/programs/cashiers-check/src/lib.rs :eyes:

**Comment by alexiskattan** [comment]
@CriesofCarrots thank you! It worked! Ah didn't realize that `TOKEN_PROGRAM_ID` was a const and not a variable.... Looks great.  Thank you for all the help on this!

**Comment by alexiskattan** [comment]
@t-nelson Trying to wrap my head around this.  So I deploy the Cashier Check contract for "Token XYZ".  The Buyer then deposits X tokens into the contract along with the Seller's address.  The Seller than pings the Contract every few seconds to see if there are tokens in the contract for them.  If there are tokens for the Seller, they cash the contract and are able to see who deposited them.  The seller can then send the eBook to that address.  Is that right?

**Comment by alexiskattan** [comment]
thank you for the help on this @CriesofCarrots.  circling back on previous question about calling `create-account` twice.

when I switch from CLI to the JS library I'm getting some behavior that I either don't get or is a bug.  Here is the minimum program.  Basically I get a token then try and call create account on it twice for a given address but it returns two different values.

```
import token3 from  '@solana/spl-token';
import web3 from '@solana/web3.js';

const TOKEN_ID = new web3.PublicKey(
    'C2GcearUU9MdByy9ngHAUBYkcxR84WvjCJWBeAu1mMvg'
)

// send tokens
const token = new token3.Token(
    connection,
    TOKEN_ID,        
    token3.TOKEN_PROGRAM_ID,
    payerAccount,
);

(async() => {
    // these two return two different values
    let token_account_1 = await token.createAccount(
        "AcVM2Kt5V92eXSDgS8bzPfYyy9hGuUM2LCzCr7vD4mfo",
    );

    let token_account_2 = await token.createAccount(
        "AcVM2Kt5V92eXSDgS8bzPfYyy9hGuUM2LCzCr7vD4mfo",
    );
   //token_account_1 does not equal token_account_2, which is what i would expect after using the CLI interface

})();
```

**Comment by CriesofCarrots** [comment]
@alexiskattan , sorry for the delay. Github isn't really the best place for technical support. However:
[Token.createAccount](https://github.com/solana-labs/solana-program-library/blob/0b3552b8926a9e4b37074a3d19d06591d47ed50a/token/js/client/token.js#L446) is the analog to `create-account AQoKYV7tYpTrFZN6P5oUufbQKAUr9mNYGe1TTJC9wajM <KEYPAIR>`, where `<KEYPAIR>` is randomly generated each time you call the method.

I think the method you actually want here is [Token.createAssociatedTokenAccount](https://github.com/solana-labs/solana-program-library/blob/0b3552b8926a9e4b37074a3d19d06591d47ed50a/token/js/client/token.js#L494)
